File: Day1/part2.py
class Group():
    def __init__(self, length=3, index_offset=1):
        self.length = length
        self.index_offset = index_offset

        self.values = []

    def addValue(self, value):
        if self.isFull() == False:
            self.values.append(value)

    def isFull(self):
        return True if len(self.values) >= self.length else False

    def indexIsInGroup(self, index):
        offsetIndex = index + self.index_offset
        if (offsetIndex % 4) != 0:
            return True
        else:
            return False

file = "input.txt"

values = []
with open(file, "r") as open_file:
    while (line := open_file.readline()):
        values.append(
            int(line.strip())
        )

increases = 0

sums = []
activeGroups = []

for i in range(len(values)):
    if len(activeGroups) < 3:
        activeGroups.append([])

    for group in activeGroups:
        group.append(values[i])

    for group in activeGroups:
        if len(group) == 3:
            sums.append(group[0]+group[1]+group[2])
            activeGroups.remove(group)


for i in range(1, len(sums)):
    if sums[i] > sums[i-1]:
        increases += 1

print(increases)

File: Java/2021/Day1/part2.java
import java.io.File;
import java.io.FileNotFoundException;
import java.util.Scanner;
import java.util.List;
import java.util.ArrayList;

class Part2 {
  public static void main(String[] args) throws FileNotFoundException {
    List<Integer> data = readFile("input.txt");

    ArrayList<ArrayList<Integer>> activeGroups = new ArrayList<ArrayList<Integer>>();

    List<Integer> sums = new ArrayList<Integer>();

    for (int i = 0; i < data.size(); i++) {
      if (activeGroups.size() < 3) {
        activeGroups.add(new ArrayList<Integer>());
      }

      for (ArrayList<Integer> group : activeGroups) {
        group.add(data.get(i));
      }

      ArrayList<ArrayList<Integer>> groupsCopy = new ArrayList<ArrayList<Integer>>(activeGroups);
      for (int j = 0; j < activeGroups.size(); j++) {
        if (activeGroups.get(j).size() >= 3) {
          ArrayList<Integer> group = activeGroups.get(j);
          int sum = group.get(0) + group.get(1) + group.get(2);

          sums.add(sum);

          groupsCopy.remove(groupsCopy.indexOf(group));
        }
      }

      activeGroups = groupsCopy;
    }


    int increases = 0;

    for (int i = 1; i < sums.size(); i++) {
      if (sums.get(i) > sums.get(i-1)) {
        increases++;
      }
    }

    System.out.println(increases);
  }

  private static List<Integer> readFile(String path) throws FileNotFoundException {
    File myFile = new File(path);
    Scanner fileScanner = new Scanner(myFile);

    List<Integer> data = new ArrayList<Integer>();

    while (fileScanner.hasNextLine()) {
      String line = fileScanner.nextLine().trim();
      data.add(Integer.parseInt(line));
    }

    return data;
  }
}
